File: hw2/main.py

import time
import numpy as np
from bfs.seq import seq_bfs
from bfs.par import par_bfs


def generate_cubic_graph(side_length):
    graph = {}
    for x in range(side_length):
        for y in range(side_length):
            for z in range(side_length):
                neighbors = []
                for dx, dy, dz in [(1, 0, 0), (-1, 0, 0), (0, 1, 0),
                                   (0, -1, 0), (0, 0, 1), (0, 0, -1)]:
                    nx, ny, nz = x + dx, y + dy, z + dz
                    if 0 <= nx < side_length and 0 <= ny < side_length and 0 <= nz < side_length:
                        neighbors.append((nx, ny, nz))
                graph[(x, y, z)] = neighbors
    return graph


def test_correctness():
    small_graph = {
        (0, 0): [(0, 1), (1, 0)],
        (0, 1): [(0, 0), (1, 1)],
        (1, 0): [(0, 0), (1, 1)],
        (1, 1): [(0, 1), (1, 0)]
    }
    start_node = (0, 0)

    result_seq = seq_bfs(small_graph, start_node)
    result_par = par_bfs(small_graph, start_node)

    # print(*result_par, *result_seq)
    assert result_seq == result_par, "Diff"

    side_length = 10
    large_graph = generate_cubic_graph(side_length)
    start_node = (0, 0, 0)
    
    result_seq = seq_bfs(large_graph, start_node)
    result_par = par_bfs(large_graph, start_node)

    assert result_seq == result_par, "Diff"

    print("Same results!")


def test():
    side_length = 10
    start_node = (0, 0, 0)
    graph = generate_cubic_graph(side_length)

    seq_times = []
    for i in range(5):
        start_time = time.time()
        seq_bfs(graph, start_node)
        elapsed_time = time.time() - start_time
        seq_times.append(elapsed_time)

    par_times = []
    for i in range(5):
        start_time = time.time()
        par_bfs(graph, start_node, num_processes=4)
        elapsed_time = time.time() - start_time
        par_times.append(elapsed_time)

    seq_avg_time = sum(seq_times) / len(seq_times)
    par_avg_time = sum(par_times) / len(par_times)

    speedup = seq_avg_time / par_avg_time if par_avg_time > 0 else float('inf')
    
    return seq_avg_time, par_avg_time, speedup

if __name__ == "__main__":
    print("Correctness...")
    test_correctness()
    print("Perfect!")

    print("\nTests:")
    seq_avg_time, par_avg_time, speedup = test()

    print(f"Av time (seq): {seq_avg_time:.4f} s")
    print(f"Av time (par): {par_avg_time:.4f} s")
    print(f"{speedup:.2f}x")

File: hw2/main/main.go
package main

import (
	"log"
	"time"
)

func generateCubicGraph(side int) Graph {
	graph := make(Graph)

	for x := 0; x < side; x++ {
		for y := 0; y < side; y++ {
			for z := 0; z < side; z++ {
				node := x*side*side + y*side + z
				neighbors := []int{}

				if x > 0 {
					neighbors = append(neighbors, (x-1)*side*side+y*side+z)
				}
				if x < side-1 {
					neighbors = append(neighbors, (x+1)*side*side+y*side+z)
				}
				if y > 0 {
					neighbors = append(neighbors, x*side*side+(y-1)*side+z)
				}
				if y < side-1 {
					neighbors = append(neighbors, x*side*side+(y+1)*side+z)
				}
				if z > 0 {
					neighbors = append(neighbors, x*side*side+y*side+(z-1))
				}
				if z < side-1 {
					neighbors = append(neighbors, x*side*side+y*side+(z+1))
				}

				graph[node] = neighbors
			}
		}
	}

	return graph
}

type Graph map[int][]int

func eq(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i, ai := range a {
		if ai != b[i] {
			return false
		}
	}
	return true
}

func test(gr Graph, times int64) {
	var s1, s2 int64
	for i := range times {
		log.Println(">> time ", i, " out of ", times)
		n := time.Now()
		r1 := SeqBFS(gr, 0, make([]int, len(gr)))
		s1 += time.Since(n).Nanoseconds()

		n = time.Now()
		r2 := ParBFS(gr, 0, make([]int, len(gr)))
		s2 += time.Since(n).Nanoseconds()

		if !eq(r1, r2) {
			panic("Different r1, r2!")
		}
		log.Printf("### %d Result avg time: seq %.3f, par %.3f, speed up: %f",
			i, float64(s1)/float64(i+1),
			float64(s2)/float64(i+1), float64(s1)/float64(s2))

	}
	log.Printf("Result avg time: \nseq: %.3f,\npar: %.3f,\nspeed up: %f",
		float64(s1)/float64(times),
		float64(s2)/float64(times),
		float64(s1)/float64(s2))

}

func main() {
	graph := Graph{
		0: {1, 3},
		1: {2, 3},
		2: {4, 5},
		3: {},
		4: {},
		5: {6},
		6: {},
	}
	test(graph, 1)

	gr := generateCubicGraph(100)
	log.Println("size of graph: ", len(gr))
	test(gr, 5)
}
